runtime/common/src/fees.rs

```diff
@@ -0,0 +1,437 @@
+            }
+        }
+
+        pub struct DealWithForeignFees;
+
+        impl OnUnbalanced<CreditOf<AccountId, Tokens>> for DealWithForeignFees
+        {
+            fn on_unbalanced(fees_and_tips: CreditOf<AccountId, Tokens>) {
+                debug_assert!(FEES_AND_TIPS_TREASURY_PERCENTAGE + FEES_AND_TIPS_BURN_PERCENTAGE == 100u32);
+                let total_percentage = (FEES_AND_TIPS_TREASURY_PERCENTAGE.saturating_add(FEES_AND_TIPS_BURN_PERCENTAGE)) as u128;
+                let fees_and_tips_value = fees_and_tips.peek().clone();
+                // Split the merged imbalance into two parts
+                let (split_for_treasury, split_for_burn) = fees_and_tips.split(
+                    fees_and_tips_value * FEES_AND_TIPS_TREASURY_PERCENTAGE as u128 / total_percentage
+                );
+                // Handle the split imbalances
+                // on_unbalanced is not implemented for other currencies than the native currency
+                // https://github.com/paritytech/substrate/blob/85415fb3a452dba12ff564e6b093048eed4c5aad/frame/treasury/src/lib.rs#L618-L627
+                // https://github.com/paritytech/substrate/blob/5ea6d95309aaccfa399c5f72e5a14a4b7c6c4ca1/frame/treasury/src/lib.rs#L490
+                let _ = <Tokens as Balanced<AccountId>>::resolve(&TreasuryPalletId::get().into_account_truncating(), split_for_treasury);
+                // Burn the remaining part
+                drop(split_for_burn);
+            }
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! impl_foreign_fees {
+    {} => {
+        use frame_support::unsigned::TransactionValidityError;
+        use orml_traits::arithmetic::One;
+        use orml_traits::arithmetic::Zero;
+        use frame_support::traits::fungibles::CreditOf;
+        use frame_support::traits::tokens::fungibles::Inspect;
+        use frame_support::traits::tokens::WithdrawReasons;
+        use frame_support::traits::tokens::BalanceConversion;
+        use frame_support::traits::ExistenceRequirement;
+        use frame_support::traits::tokens::WithdrawConsequence;
+        use sp_runtime::traits::DispatchInfoOf;
+        use sp_runtime::traits::PostDispatchInfoOf;
+        use frame_support::pallet_prelude::InvalidTransaction;
+        use frame_support::traits::tokens::fungibles::Balanced;
+        use zrml_swaps::check_arithm_rslt::CheckArithmRslt;
+        use sp_runtime::traits::Convert;
+        use pallet_asset_tx_payment::HandleCredit;
+
+        // It does foreign fees by extending transactions to include an optional `AssetId` that specifies the asset
+        // to be used for payment (defaulting to the native token on `None`). So for each transaction you can specify asset id
+        // For real ZTG you use None and for orml_tokens ZTG you use `Some(Asset::Ztg)` and for DOT you use `Some(Asset::Foreign(0))`
+
+        pub(crate) fn calculate_fee(
+            native_fee: Balance,
+            fee_factor: Balance,
+            base: Balance,
```

Comment:
It's the decimal number of a foreign currency. It is not the necessary the same base as the native currency of 10 decimal places. `BASE` is the number `10_000_000_000`.

<img width="744" alt="Screenshot 2023-07-14 at 08 38 42" src="https://github.com/zeitgeistpm/zeitgeist/assets/33418719/c3d696c9-d8a8-496a-90ae-b094097c9a2a">

Reply:
Okay, we have to be careful here:

- All foreign assets are represented with 10 decimal places on our system, and this will always be the case. Foreign assets with more decimal places (resp. fewer) will be truncated (resp. padded). This makes the `base` parameter unnecessary.
- Your code assumes that the native currency and the foreign currency have the same number of decimal places, specified by the `base` parameter. So with the current implementation, if you did use a foreign currency with `decimal` decimal places and `base = 10 ** decimals` in `calculate_fee(native_fee, fee_factor, base)`, then `bmul` will interpret the `native_fee` as a fixed point number with `decimal` decimal places, which is incorrect unless `decimals` happens to be `10`. You can verify this in your tests. If you change `decimal` in `withdraws_correct_dot_foreign_asset_fee` to `12`, you would expect `7_156_026_000` (which represents `0.007156026` as a fixed point number with 12 decimals, the same as the original `71_560_260` with respect to 10 decimals). But instead it gives you `715603`.

Reply:
If we had a currency with 12 decimal places, we also would have an own fee factor for this currency, which would change the result.

Reply:
So this `native = native * base / BASE` would have been the solution to get from the native denominated in ZTG (10 decimal places) to the foreign currency base of for example 12 decimal places (maybe KSM), right? I see now. Harald told me that fee_factor is calculated based on the foreign currency base, but with his PR we go completely to 10 decimal places as you described for all foreign currencies (truncating if necessary).

Reply:
https://github.com/zeitgeistpm/zeitgeist/pull/1022/commits/bd4840837175280b07187331093ef3f502759a79